services/httppost/service.go

```diff
@@ -0,0 +1,253 @@
+package httppost
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"sync"
+
+	"github.com/influxdata/kapacitor/alert"
+	"github.com/influxdata/kapacitor/bufpool"
+)
+
+// Only one of name and url should be non-empty
+type Endpoint struct {
+	mu      sync.RWMutex
+	url     string
+	headers map[string]string
+	closed  bool
+}
+
+func NewEndpoint(url string, headers map[string]string) *Endpoint {
+	return &Endpoint{
+		url:     url,
+		headers: headers,
+	}
+}
+func (e *Endpoint) Close() {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	e.closed = true
+	return
+}
+
+func (e *Endpoint) SetURL(url string) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	e.url = url
+}
+
+func (e *Endpoint) SetHeaders(headers map[string]string) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	e.headers = headers
+}
+
+func (e *Endpoint) NewHTTPRequest(body io.Reader) (req *http.Request, err error) {
+	e.mu.RLock()
+	defer e.mu.RUnlock()
+	if e.closed {
+		return nil, errors.New("endpoint was closed")
+	}
+
+	req, err = http.NewRequest("POST", e.url, body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create POST request: %v", err)
+	}
+
+	for k, v := range e.headers {
+		req.Header.Add(k, v)
+	}
+
+	return req, nil
+}
+
+type Service struct {
+	mu        sync.RWMutex
+	endpoints map[string]*Endpoint
+	logger    *log.Logger
+}
+
+func NewService(c Configs, l *log.Logger) *Service {
+	s := &Service{
+		logger:    l,
+		endpoints: c.index(),
+	}
+	return s
+}
+
+func (s *Service) Endpoint(name string) (*Endpoint, bool) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	e, ok := s.endpoints[name]
+	return e, ok
+}
+
+func (s *Service) Update(newConfigs []interface{}) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	endpointSet := map[string]bool{}
+
+	for _, nc := range newConfigs {
+		if c, ok := nc.(Config); ok {
+			if err := c.Validate(); err != nil {
+				return err
+			}
+			e, ok := s.endpoints[c.Endpoint]
+			if !ok {
+				s.endpoints[c.Endpoint] = NewEndpoint(c.URL, c.Headers)
+				continue
+			}
+			e.SetURL(c.URL)
```

Comment:
I think we want these Set calls to be atomic together. Meaning there is a slight chance that the endpoint is used right after the URL is modified but before the Headers are which could cause issues. Maybe we should just have an Update method that takes a Config? Lots of ways to do it, but all updates should happen with one lock/unlock of the mutex.